### PR TITLE
Task Priority Escalation

### PR DESCRIPTION
**Pull request by rokhinip**
This change has 3 major parts to it: 

1.  Change ActiveTaskStatus layout in order to be able to track the identity
    of the thread that is executing the task as part of the ActiveTaskStatus

2.  Allow concurrent modifications to the ActiveTaskStatus while another
    thread has the task status record lock.
    
    Today, if a thread is holding the StatusRecordLock, then no other
    modification of the task status is possible - including a thread
    starting to execute the task or stopping execution of the task.
    However, the TaskStatusRecordLock is really about protecting the linked
    list that is maintained in the ActiveTaskStatus. As such, other
    operations which don't need to look at that linked list of us records
    really shouldn't have to block on the StatusRecordLock.
    
    This change allows for concurrent modification of the ActiveTaskStatus while
    the TaskStatusRecordLock is held. In particular, a task can cancelled,
    escalated, start and stop running, all while another thread is holding onto
    the task's StatusRecordLock. In the event of cancellation and
    escalation, the task's StatusRecordLock must be acquired in order to propagate
    cancellation and escalation to its child tasks but it is not needed to cancel
    or escalate the task itself.

3.  Implement task priority escalation on Apple platforms
    
    A task can be in one of 4 states over its lifetime:
    
        (a) suspended
        (b) enqueued
        (c) running
        (d) completed
    
    This change provides priority inversion avoidance support if a task gets
    escalated when it is in state (a), (c), (d).

Rdar: rdar://76127624

### CODE REVIEW

**Comment by rokhinip** [comment]
I recommend reviewing the PR commit by commit.

File: include/swift/Runtime/Concurrency.h
Comment (rokhinip) [inline include/swift/Runtime/Concurrency.h:48]: Is there a more robust/explicit way to determine platforms which support 128 bit DCAS?

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (mikeash) [inline stdlib/public/Concurrency/TaskStatus.cpp:588]: `offsetof` or casting to `ActiveTaskStatus *` and taking the address of the field would be a bit clearer here.

**Review by mikeash — APPROVED** [review]

File: include/swift/Runtime/Concurrency.h
Comment (mikeash) [reply include/swift/Runtime/Concurrency.h:48]: Once we move to C++17, we could probably figure it out with `std::atomic<T>::is_always_lock_free`, but we can't use that yet.

**Comment by rokhinip** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [inline stdlib/public/Concurrency/TaskPrivate.h:290]: I don't know where the right funnel point in code for when a task is enqueued - whether it is on an actor or a global pool. If we had such a funnel point, it would make it easier to

(a) Set the enqueued bit on the task state
(b) Copy over the maxPriority from active task status to the job field. Right now, this operation happens in a bunch of places and could be consolidated to one funnel point.

All the entry points that I have found for enqueuing work onto an executor only take Job structures though. Should I provide overloads for all of the `swift_task_enqueue` functions such that they task Task objects?

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rokhinip) [inline stdlib/public/Concurrency/TaskStatus.cpp:279]: I don't see why this has to be a release barrier, it seems like

(a) we're simply updating the ActiveTaskStatus's TaskStatusRecord pointer with what was the previous head pointer and all the state we want to publish is within that atomic.

(b) We're not making any other writes to the new record we're publishing, such that it needs to happen before this newStatus is visible.

That's why I changed this from a release barrier to a relaxed. Let me know if I'm missing something here.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (mikeash) [reply stdlib/public/Concurrency/TaskPrivate.h:290]: I think you'll want to add a method on `Job` that you call from the enqueue functions. Then that method can do a dynamic cast and call a Task method, like, `if (auto task = dyn_cast<AsyncTask>(this)) task->handleEnqueueing()`, to do task-specific stuff.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (mikeash) [reply stdlib/public/Concurrency/TaskStatus.cpp:279]: Your analysis makes sense to me. Maybe this was written with the general idea of "we're publishing a pointer, better make sure the contents are published first" without realizing that the contents aren't being changed here.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskPrivate.h:232]: Yeah, this kind of "it's just like Dispatch" comment isn't really okay here, because other maintainers of this code (like me!) don't have any idea what it means.

In general, I would strongly prefer that we not widely introduce Dispatch jargon like "drainer" into the Swift runtime except when specifically necessary to talk about an interaction with Dispatch.  Nothing "drains" a task.

I'm not sure I understand the value of tracking whether we're currently enqueued given that we're not tracking the identity of the queue (yet).

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rjmccall) [reply stdlib/public/Concurrency/TaskPrivate.h:290]: Where do we actually want to do the record-keeping associated with enqueueing a task?  My intuition is that in most cases we'll want to take care of this in the caller of `swift_task_enqueue` because we'll have to say that the task is no longer waiting on whatever it was waiting on, but maybe that's not true and it should be part of `swift_task_enqueue`.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskStatus.cpp:469]: For the reasons you mention below, you can return if there aren't any records on `oldStatus`.

We've lost an optimization to not modify the atomic multiple times when we do need to install the lock, but I guess that's not particularly important.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rjmccall) [reply stdlib/public/Concurrency/TaskStatus.cpp:279]: Yeah, I think this is fine.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskStatus.cpp:582]: Trailing sentence

Is "debounce" Dispatch jargon for this?  As a general concept, debouncing means combining multiple events into one, which is not what this is about — this is about making sure that the thread priority isn't changed for stale reasons.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskStatus.cpp:615]: And just like with cancellation:
- we can therefore avoid getting the lock when the previous atomic use saw that there were no records
- this comment is misleading

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskStatus.cpp:478]: I think this comment is a little misleading.  It doesn't actually matter for our local purposes that records can be added between when we set the cancelled bit and when we acquire the lock, because in any event we're going to visit all the records that are in place when we acquire the lock.  `updateNewChild...` isn't important for *those* records, it's important for the records being concurrently added that we *don't* see when we acquire the lock, i.e. the records that from our perspective are added strictly after us.

File: include/swift/Reflection/RuntimeInternals.h
Comment (mikeash) [reply include/swift/Reflection/RuntimeInternals.h:96]: Yes, this is used by Remote Mirror to peek at this stuff from outside the process. We can easily make this different for 32/64-bit, but we can't easily conditionalize it on `SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION`, so making the layouts the same regardless of `SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION` would be best if that's possible. Feel free to hit me up offline if you want to chat about specifics.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:290]: Let me take a stab at doing it in each caller and seeing how that looks. I think you might be right that we might want to do something different based on the different calling contexts to `swift_task_enqueue` and its friends.

Since we have multiple flavors of enqueue functions as well, I was hoping to avoid having to hunt down each caller.

File: include/swift/Runtime/Concurrency.h
Comment (rokhinip) [reply include/swift/Runtime/Concurrency.h:57]: Ah I figured that this ought to sit next to the other defines above which are also about how the runtime executes - using cooperative executor etc. What would be a more appropriate place to move this?

File: include/swift/Reflection/RuntimeInternals.h
Comment (rokhinip) [reply include/swift/Reflection/RuntimeInternals.h:96]: I actually didn't change the total size of ActiveTaskStatus - it is still 2 words. The internal bits have been split up further.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:201]: Understood, I might tweak your suggestion and omit information about the syscall interface on darwin because that is really more about documentation about dispatch's interfaces and shouldn't live here. Not to mention that it can change in the future.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:232]: Yea I can work on better names and comments. I just started with what felt natural to me, in order to things bootstrapped and working end-to-end.

> I'm not sure I understand the value of tracking whether we're currently enqueued given that we're not tracking the identity of the queue (yet).

a) I think It is helpful for us to understand whether the task is in the state of suspended, runnable or running and therefore what kinds of action we need to take for priority escalation. We currently don't do anything for the `runnable` case (ie no stealing) but it is helpful bookkeeping to start having.

b) It also helps us identity whether the Job fields should make sense or not since they really only are of value in enqueued jobs when inspecting the structure from outside.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [inline stdlib/public/Concurrency/TaskPrivate.h:358]: TODO self: Indent seems off here.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskStatus.cpp:478]: Excellent point. Will update.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskStatus.cpp:469]: Good point will add a short-circuit.

You're right that we are writing to the same atomic multiple times ...but... I do hope to be able to make a future change which allows us to avoid a full double-wide CAS when we really only care about modifying the stuff in the top word like in this loop. We should only need the double-wide CAS when we are taking the status record lock.  That should ease some register pressure at least.

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskStatus.cpp:582]: Hmmm, let me check with folks offline.

Comment (rokhinip) [reply stdlib/public/Concurrency/TaskStatus.cpp:582]: Looking at the linux futex docs (https://man7.org/linux/man-pages/man2/futex.2.html), they don't seem to have a singular word or phrase that really describes what the operation is trying to do: 

```
When executing a futex operation that requests to block a thread,
the kernel will block only if the futex word has the value that
the calling thread supplied (as one of the arguments of the
futex() call) as the expected value of the futex word.  The
loading of the futex word's value, the comparison of that value
with the expected value, and the actual blocking will happen
atomically and will be totally ordered with respect to concurrent
operations performed by other threads on the same futex word.
Thus, the futex word is used to connect the synchronization in
user space with the implementation of blocking by the kernel.
Analogously to an atomic compare-and-exchange operation that
potentially changes shared memory, blocking via a futex is an
atomic compare-and-block operation.
```

@rjmccall I'm open to suggestions of how else to succintly name the operation described above.

Alternatively, I'm inclined to just remove this comment altogether since it's really about documenting what dispatch is doing now and might want to do in the future. I put it here cause I was co-developing the dispatch interface at the same time and so the lines were blurry to me.

File: include/swift/Reflection/RuntimeInternals.h
Comment (rokhinip) [reply include/swift/Reflection/RuntimeInternals.h:96]: Mike and I worked on fixing this offline. Fixup incoming.

**Comment by rokhinip** [comment]
@swift-ci please test

File: include/swift/Reflection/ReflectionContext.h
Comment (rokhinip) [inline include/swift/Reflection/ReflectionContext.h:104]: @mikeash, our changes to RuntimeInternals.h caused a bunch of unhappiness here and so I took a stab at fixing it, including stealing some logic from your PR #41160.

Would appreciate a look-over.

**Comment by rokhinip** [comment]
@swift-ci Please test Windows platform

File: include/swift/Reflection/ReflectionContext.h
Comment (mikeash) [reply include/swift/Reflection/ReflectionContext.h:104]: Looks good! We'll get some conflicts and need some more fixes once one of us merges, but this is good for the moment.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:320]: I've taken a stab at fixing this and I think this actually really cleans up the state machine of identifying the various states a task can be in. Feedback appreciated!

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
@swift-ci Please test Windows platform

File: include/swift/Runtime/Concurrency.h
Comment (rokhinip) [inline include/swift/Runtime/Concurrency.h:637]: @rjmccall, with the changes I made to provide a new `flagAsEnqueuedOnExecutor` on the task, which does the necessary bookkeeping and enqueuing of the task, these unit tests started failing since they were directly calling `swift_task_enqueue` without doing the right bookkeeping ahead of time.

I therefore had to introduce a new exported function here, for the sole purpose of being able to get the unit tests working and exercising the logic here. It is not needed for the actual logic to work. And on that note, `swift_task_enqueue` and `swift_task_enqueueGlobal` should also not to be exported since they are only internal to the runtime library.

How do you want to proceed here?

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (mikeash) [inline stdlib/public/Concurrency/TaskStatus.cpp:586]: This can be simplified to `executionLock = &taskStatus->ExecutionLock;`.

**Review by mikeash — APPROVED** [review]

**Comment by rokhinip** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskStatus.cpp:586]: Actually don't think so since ExecutionLock is a private member of the ActiveTaskStatus. Maybe it shouldn't be though? Keeping this as is for now.

**Comment by rokhinip** [comment]
@swift-ci Please test macOS platform

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
@swift-ci  Please test macOS platform

File: stdlib/public/Concurrency/TaskStatus.cpp
Comment (mikeash) [reply stdlib/public/Concurrency/TaskStatus.cpp:586]: Ah, yeah. Seems like it probably shouldn't be private given this. Or there could be a get method that returns a reference to it. That can be a followup though.

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
@swift-ci please test

File: include/swift/Runtime/Concurrency.h
Comment (rjmccall) [reply include/swift/Runtime/Concurrency.h:637]: Hmm.  I hadn't really thought about the public entrypoints.  They're not usually directly useful, but:

- We do have a way to create a job in an unscheduled state, and we ought to have *some* way to schedule the task afterwards.  That could be a new operation.  I guess it depends on whether you actually found it useful to do the bookkeeping in the caller, or whether it's necessary to touch the status during enqueue regardless?  I need to look at the new patch.

- We also ought to continue to provide some way to hook the scheduling process, in case we need that for future compatibility hackery.  So at any rate, we shouldn't remove the ability to replace these.

File: stdlib/public/Concurrency/Actor.cpp
Comment (rjmccall) [inline stdlib/public/Concurrency/Actor.cpp:1621]: From the names, I'd expect these methods to just be bookkeeping.  If they're doing the actual work of enqueuing the task, they should probably have a verb as their name, probably `enqueue`.

File: stdlib/public/Concurrency/Debug.h
Comment (rjmccall) [inline stdlib/public/Concurrency/Debug.h:50]: Ah, yeah, this is a good way to deal with this.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskPrivate.h:232]: This alignment probably needs to be 16 if we're using priority escalation on a 32-bit system, like we presumably want to do on `arm64_32`.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rjmccall) [inline stdlib/public/Concurrency/TaskPrivate.h:231]: Thanks, this comment looks great.

Well, except this last paragraph is confusing.  We're not actually doing partial atomic updates, right?  It's just the partial atomic read from the kernel.

File: include/swift/Reflection/RuntimeInternals.h
Comment (rjmccall) [inline include/swift/Reflection/RuntimeInternals.h:99]: This doesn't look right; it should be:

```
  uint32_t Flags;
  uint32_t DrainLock;
  typename Runtime::StoredPointer StatusRecords;
```

Right?

File: include/swift/Reflection/RuntimeInternals.h
Comment (rjmccall) [inline include/swift/Reflection/RuntimeInternals.h:105]: This also looks wrong.

File: include/swift/Runtime/Concurrency.h
Comment (rjmccall) [reply include/swift/Runtime/Concurrency.h:637]: Is there a way to set up a created-but-not-scheduled task's status so that enqueuing it just works, or do we need to add specific new API for this?

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:231]: Currently we are doing full double wide atomic updates always but we don't always need to do that.

We can do relaxed loads on the top word and only do the actual double-wide atomics if the isStatusRecordLockedBit is set and we care to take the record lock. If we don't need the lock (like for instance, for cancellation of a task that has no children), then we don't need to double-wide atomics. It's a nice micro-optimization that is available to us and will ease register pressure in a lot of points.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskPrivate.h:232]: Yes you're right and I realized this yesterday while I was testing for arm64_32. It has larger implications on the layout of PrivateStorage and so in an attempt to not cram more stuff into a single PR that is big enough as is, I decided to change this PR to only enable `SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION` when `SWIFT_POINTER_IS_8_BYTES`. 

I will be making a follow up PR to handle arm64_32.

File: stdlib/public/Concurrency/Actor.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/Actor.cpp:1621]: Yea I was conflicted as well - it felt really nice to keep it in line with the other `flagAs` functions since they nicely encapsulate all the state changes of a task. I can make this clean-up in a follow on PR.

File: include/swift/Reflection/RuntimeInternals.h
Comment (rokhinip) [reply include/swift/Reflection/RuntimeInternals.h:99]: Yes this is kind of a gross-ish hack that @mikeash and I came up with. It is meant to reflect the following layouts which are possible in 32 and 64 bit case of supporting priority escalation:

32 bit supporting priority escalation - ie arm64_32: 

```
uint32_t flags
uint32_t drainLock
uint32_t unused
typename Runtime::StoredPointer statusRecords
```

64 bit supporting priority escalation: 

```
uint32_t flags
uint32_t drainLock
typename Runtime::StoredPointer statusRecords
```

File: include/swift/Reflection/RuntimeInternals.h
Comment (rokhinip) [reply include/swift/Reflection/RuntimeInternals.h:105]: Same hack logic as above: 

32 bit without priority escalation:

```
uint32_t flags
typename Runtime::StoredPointer statusRecords
```
64 bit without priority escalation:
```
uint32_t flags
uint32_t unused
typename Runtime::StoredPointer statusRecords
```

File: include/swift/Runtime/Concurrency.h
Comment (rokhinip) [reply include/swift/Runtime/Concurrency.h:637]: > We do have a way to create a job in an unscheduled state, and we ought to have some way to schedule the task afterwards. That could be a new operation. I guess it depends on whether you actually found it useful to do the bookkeeping in the caller, or whether it's necessary to touch the status during enqueue regardless? I need to look at the new patch.

Is "create a job but don't enqueue it" currently supported in the language? I agree that it is a useful thing to support in the future. In my current patch, `flagAsEnqueuedOnExecutor` handles 2 possible transitions of a task to the enqueued state: `suspended->enqueued` and `running->enqueued`. When a task is created but not scheduled, that falls under the first transition that is handled in `flagAsEnqueuedOnExecutor`.

I think both codepaths - whether the task is coming from suspended or from running - will need to touch the status to at least set the enqueued bit on the task and sometimes need to touch the status in the caller. We may be able to refactor this into `flagAsEnqueuedFromRunning()` and `flagAsEnqueuedFromSuspended` in order to keep all the atomic operations on the status together. 

> Is there a way to set up a created-but-not-scheduled task's status so that enqueuing it just works, or do we need to add specific new API for this?

Yes and that's what we do in `swift_task_create_commonImpl`. The task is created earlier and then we call flagAsEnqueuedOnExecutor and enqueueing just works.

Comment (rjmccall) [reply include/swift/Runtime/Concurrency.h:637]: > Is "create a job but don't enqueue it" currently supported in the language?

There's not a way to do it, no, but that's not the point; if we add an optimization / new feature that would use this facility, the question is whether it can back-deploy.  It's currently tentatively supported by the ABI; if that support isn't really enough to implement a reasonable future feature, so be it, we can say that that feature doesn't back-deploy; but it'd be nice to at least think about this.

> Yes and that's what we do in swift_task_create_commonImpl

No, I mean in the path where `taskCreateFlags.enqueueJob()` is not true.

File: stdlib/public/Concurrency/Actor.cpp
Comment (rjmccall) [reply stdlib/public/Concurrency/Actor.cpp:1621]: Okay.  Originally these methods were all just bookkeeping, so since they've now all become the actual way to do things, they should all be renamed.

File: stdlib/public/Concurrency/TaskPrivate.h
Comment (rjmccall) [reply stdlib/public/Concurrency/TaskPrivate.h:232]: Okay, that's fine.

File: include/swift/Reflection/RuntimeInternals.h
Comment (rjmccall) [reply include/swift/Reflection/RuntimeInternals.h:99]: Oh, I misread the code, sorry.

File: stdlib/public/Concurrency/Actor.cpp
Comment (rokhinip) [reply stdlib/public/Concurrency/Actor.cpp:1621]: I will do that in my next PR.

File: include/swift/Runtime/Concurrency.h
Comment (rokhinip) [reply include/swift/Runtime/Concurrency.h:637]: Ah I see where you're going with this - I didn't consider back deployment. 

There is currently nothing special in the status of a task prior to enqueueing - it just needs to not already have an enqueued bit. 

Assuming we consider "create a schedule-able task to enqueue later" as being in 3 parts: 
1. create a suspended task
2. (Optionally) Update a suspended task with any state changes if needed
3. Enqueue the suspended task when it is ready

All 3 parts are doable with the hook points we have today:
(1) with `taskCreateFlags.enqueueJob() == false`, 
(2) This is handled by the various operations that may need to touch the task like escalation and cancellation
(3) We can enqueue the suspended task with `flagAsEnqueueOnExecutor`. 

I don't think we need new API for it.

**Pull request merged**
